src/main/java/org/commonjava/cdi/util/weft/LoadSensitivePoolWeftExecutorService.java

```diff
@@ -0,0 +1,261 @@
+/**
+ * Copyright (C) 2013 Red Hat, Inc. (jdcasey@commonjava.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.cdi.util.weft;
+
+import org.commonjava.cdi.util.weft.exception.PoolOverloadException;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class LoadSensitivePoolWeftExecutorService
+        implements WeftExecutorService, ScheduledExecutorService
+{
+    private PoolWeftExecutorService delegate;
+
+    public LoadSensitivePoolWeftExecutorService( PoolWeftExecutorService delegate )
+    {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public String getName()
+    {
+        return delegate.getName();
+    }
+
+    @Override
+    public boolean isHealthy()
+    {
+        return delegate.isHealthy();
+    }
+
+    @Override
+    public double getLoadFactor()
+    {
+        return delegate.getLoadFactor();
+    }
+
+    @Override
+    public int getCurrentLoad()
+    {
+        return delegate.getCurrentLoad();
+    }
+
+    @Override
+    public Integer getThreadCount()
+    {
+        return delegate.getThreadCount();
+    }
+
+    @Override
+    public void shutdown()
+    {
+        delegate.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow()
+    {
+        return delegate.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown()
+    {
+        return delegate.isShutdown();
+    }
+
+    @Override
+    public boolean isTerminated()
+    {
+        return delegate.isTerminated();
+    }
+
+    @Override
+    public boolean awaitTermination( long l, TimeUnit timeUnit )
+            throws InterruptedException
+    {
+        return delegate.awaitTermination( l, timeUnit );
+    }
+
+    @Override
+    public <T> Future<T> submit( Callable<T> callable )
+    {
+        if ( !isHealthy() )
+        {
+            throw new PoolOverloadException();
+        }
+
+        return delegate.submit( callable );
+    }
+
+    @Override
+    public <T> Future<T> submit( Runnable runnable, T t )
+    {
+        if ( !isHealthy() )
+        {
+            throw new PoolOverloadException();
```

Comment:
We should use the pool name here, and maybe give stats like the load factor and pool size here in the constructor. That would help with advice the caller might pass back to its upstream users.

Reply:
OK